### DISCUSSION TITLE
[@types/d3] - D3 v7

### DISCUSSION THREAD

**Discussion by curran** [DefinitelyTyped/DefinitelyTyped#66320] [Issues with a @types package]
It appears that types for D3 only go up to version 6

https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/d3

D3 version 7 is the latest version. I suppose a new additional folder is needed there for V7.

Here's the changelog for v7 https://github.com/d3/d3/blob/main/CHANGES.md#changes-in-d3-70

[a comment by an automated account is not shown]

**Comment by curran** [comment]
Here's a detailed idea about how the types for `csvParse` may be improved, from @pcclarke via the [D3 Slack](https://d3js.slack.com/join/shared_invite/zt-1neihq96a-xkXVPXYOmKg8ou7DL3kr7g):

I’m trying to use d3 with TypeScript, and I noticed that the items returned from csvParse are typed as string | undefined , as they are set by [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/d3-dsv/index.d.ts#L20) to DSVRowStrings. This is a bit annoying as I have to then handle possible undefined values. But when I look at d3 DSV, I can’t see anywhere that would result in a single item being undefined. I’ve tried a number of CSVs with csvParse and I can’t get it to put undefined values in a row. As far as I can tell, by default (as in without passing in a conversion function) empty fields should become empty strings.
Am I missing somewhere that csvParse could return rows containing undefined? Otherwise, it seems safe to cast to string and not worry about the possibility of undefined values.

**Reply by Methuselah96** [reply]
Feel free to create a PR with this change and I cant take a look at it more closely.

**Comment by curran** [comment]
Oh actually maybe the sub-packages are the place to update - https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/d3-dsv

**Reply by Methuselah96** [reply]
Yes, the sub-packages are where the types are, d3 just re-exports all the sub-packages.

**Comment by Methuselah96** [comment]
The types go up to version 7, it's just that the most recent major version doesn't need a subfolder. You can see that [the latest version of `@types/d3` on npm is 7.4.0](https://www.npmjs.com/package/@types/d3).

**Reply by pcclarke** [reply]
Was that because the changes didn't conflict or were too minor? What determines when a new subfolder is created?

Relating back to my question, what would be the best way to handle that: create an issue, or a pull request?

**Reply by Methuselah96** [reply]
A new subfolder is created for each old major version. So when we go from v6 to v7, we create a folder for v6 and copy the types into the folder. The types at the top-level that are not in a subfolder are for the latest major version.

I'm sorry, I don't know what you're original question is, this seems like the first question you've asked on this discussion. Issuing a PR is usually the best strategy, I don't have time to fix issues.
